cmd/policy-main.go

```diff
@@ -206,72 +255,151 @@ func doGetAccessRules(targetURL string) (r map[string]string, err *probe.Error)
 	return clnt.GetAccessRules()
 }
 
-func mainPolicy(ctx *cli.Context) error {
-	// Set global flags from context.
-	setGlobalsFromContext(ctx)
+// Run policy list command
+func runPolicyListCmd(ctx *cli.Context) {
+	targetURL := ctx.Args().Last()
+	policies, err := doGetAccessRules(targetURL)
+	if err != nil {
+		switch err.ToGoError().(type) {
+		case APINotImplemented:
+			fatalIf(err.Trace(), "Unable to list policies of a non S3 url ‘"+targetURL+"’.")
+		default:
+			fatalIf(err.Trace(targetURL), "Unable to list policies of target ‘"+targetURL+"’.")
+		}
+	}
+	for k, v := range policies {
+		printMsg(policyRules{Resource: k, Allow: v})
+	}
+}
 
-	// check 'policy' cli arguments.
-	checkPolicySyntax(ctx)
+// Run policy links command
+func runPolicyLinksCmd(ctx *cli.Context) {
+	// Get alias/bucket/prefix argument
+	targetURL := ctx.Args().Last()
 
-	// Additional command speific theme customization.
-	console.SetColor("Policy", color.New(color.FgGreen, color.Bold))
+	// Fetch all policies associated to the passed url
+	policies, err := doGetAccessRules(targetURL)
+	if err != nil {
+		switch err.ToGoError().(type) {
+		case APINotImplemented:
+			fatalIf(err.Trace(), "Unable to list policies of a non S3 url ‘"+targetURL+"’.")
+		default:
+			fatalIf(err.Trace(targetURL), "Unable to list policies of target ‘"+targetURL+"’.")
+		}
+	}
 
-	if ctx.Args().First() == "list" {
+	// Extract alias from the passed argument, we'll need it to
+	// construct new pathes to list public objects
+	alias, path := url2Alias(targetURL)
+
+	isRecursive := ctx.Bool("recursive")
+	isIncomplete := false
+
+	// Iterate over policy rules to fetch public urls, then search
+	// for objects under those urls
+	for k, v := range policies {
+		// Trim the asterisk in policy rules
+		policyPath := strings.TrimSuffix(k, "*")
+		// Check if current policy prefix is related to the url passed by the user
+		if !strings.HasPrefix(policyPath, path) {
+			continue
+		}
+		// Check if the found policy has read permission
+		perm := stringToAccessPerm(v)
+		if perm != accessDownload && perm != accessPublic {
+			continue
+		}
+		// Construct the new path to search for public objects
+		newURL := alias + "/" + policyPath
+		clnt, err := newClient(newURL)
+		fatalIf(err.Trace(newURL), "Unable to initialize target ‘"+targetURL+"’.")
+		// Search for public objects
+		for content := range clnt.List(isRecursive, isIncomplete, DirFirst) {
+			if content.Err != nil {
+				errorIf(content.Err.Trace(clnt.GetURL().String()), "Unable to list folder.")
+				continue
+			}
+			// Construct the message to be displayed to the user
+			msg := policyLinksMessage{}
+			msg.Status = "success"
+			// Convert any os specific delimiters to "/".
+			msg.FullURL = filepath.ToSlash(content.URL.String())
+			if content.Type.IsDir() {
+				msg.Type = "folder"
```

Comment:
We can skip the 'CommonPrefix' since that is not useful in downloading the object itself.